Require comment content on Create & type check on Update

rachelbaker opened the pull request:
In `wp_handle_comment_submission()` core requires new comments to have content, we should match that behavior.  Comments can be edited to have empty content (if the user has `edit_comment` permission), but we should still check that at least the comment content is a string.

Included in this PR are the following functionality changes:
- Allow the Comment content to be provided for create or update requests via the `$request['content']` OR `$request['content']['raw']` parameters, matching logic from the Posts Controller. (See #2758)
- Sanitize the comment_content with `wp_filter_kses()` which matches WP Core. **EXCEPTION**: Core sanitizes the comment_content for users with the `unfiltered_html` capability with `wp_filter_kses_post()`. See ticket for discussion #2787.
- Return a WP_Error if the Comment content is empty on create.
- Return a WP_Error if the Comment content is an empty string `''` on create.
- Return a WP_Error if the Comment content is empty on update. **EXCEPTION**: when an update request [_only_ includes that status parameter](https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-comments-controller.php#L452), we do not require the Comment content parameter.

Fixes #2758 and #2284.
Addresses PART OF #2049

rachelbaker committed:
Require comment content on create and update to match core comment handling

rachelbaker committed:
Add unit tests for missing comment content

rachelbaker committed:
Allow comment content to be empty on update, but if set confirm it is a string

rachelbaker committed:
Remove extraneous content params added to a select update comment unit tests.

Originally added in 3197f8ea674c134fc6b142eb2dd005cd75c72396

rachelbaker committed:
Merge branch 'develop' into require-comment-content

rmccue commented on tests/test-rest-comments-controller.php:
```diff
@@ -674,6 +674,25 @@ public function test_create_item_invalid_date() {
+			'post'         => $this->post_id,
+			'author_name'  => 'Reverend Lovejoy',
+			'author_email' => 'lovejoy@example.com',
+			'author_url'   => 'http://timothylovejoy.jr',
```
👏

joehoyle commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -432,6 +436,10 @@ public function update_item( $request ) {
 		}
 
+		if ( isset( $request['content'] ) && ! is_string( $request['content'] ) ) {
+			return new WP_Error( 'rest_comment_content_invalid', __( 'Invalid comment content.' ), array( 'status' => 400 ) );
```
We are using `wp_filter_post_kses` for the `sanitize_callback` on this, so I'm not sure how this would fail?

rachelbaker replied:
I didn't realize that. Good catch.

rachelbaker replied:
@joehoyle Where are you seeing the `wp_filter_post_kses` callback being set for update requests?

rachelbaker replied:
NVM, I found it.

joehoyle commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -318,6 +318,10 @@ public function create_item( $request ) {
 		}
 
+		if ( empty( $request['content'] ) || ! is_string( $request['content'] ) ) {
+			return new WP_Error( 'rest_comment_content_required', __( 'Missing comment content.' ), array( 'status' => 400 ) );
```
Can we not just make this a `required` param in the schema here: https://github.com/WP-API/WP-API/blob/f73f3e8f7df3bebfc6e6c2476a4d6b2f1f2ec136/lib/endpoints/class-wp-rest-comments-controller.php#L854?

rachelbaker replied:
Well in the schema `content` is an object, like title and content is in the Posts Controller....so how would we adapt the schema to handle the create/update params that are not objects?

rachelbaker committed:
Match content handling logic from the Posts Controller and Core when sanitizing comment_content with kses

- Allow the `comment_content` to be set in a request either by `content` or `content['raw']`
- Sanitize the `comment_content` before saving to the database with `wp_filter_kses()`
- Users with the `unfiltered_html` capability will have their comment_content sanitized with `wp_filter_post_kses()`.

rachelbaker commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -728,8 +734,20 @@ protected function prepare_status_response( $comment_approved ) {
 
-		if ( isset( $request['content'] ) ) {
-			$prepared_comment['comment_content'] = $request['content'];
+		if ( isset( $request['content'] ) && is_string( $request['content'] ) ) {
```
See similar logic in the Posts Controller: https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-posts-controller.php#L782

rachelbaker committed:
Rework and attempt to simplify the logic for handling comment_content

- Reorder and remove extra string checking on the `content` and `content.raw` properties
- Return WP_Error for comment_content that is an empty string (matching `wp_handle_comment_submission()`)
- Use the schema for `wp_filter_kses_post()` sanitization
- Still add `wp_filter_kses()` sanitization for users without the unfiltered_html capability

rachelbaker commented:
Ugghh.. ran into a problem #2758

rachelbaker committed:
Only prevent creating comments with an empty string for the content

rachelbaker committed:
Handle special case where a comment can be updated and only change the status.

rachelbaker committed:
Remove the sanitization callback from the schema and add manually for comment_content.

See https://github.com/WP-API/WP-API/issues/2758

rachelbaker committed:
Update unit tests

rachelbaker committed:
Add missing required `content` key to registered fields tests

rachelbaker committed:
Correct bad logic in comment content.raw tests

rachelbaker committed:
Change conditional check on the existence of $prepared_comment['comment_content'] to use `isset()`

rachelbaker committed:
Attempt to fix update comment content test for PHP 5.2 and 5.3

rachelbaker committed:
Maybe Disco Stu has started self-promoting

rachelbaker committed:
I hate PHP 5.3  and lower

rmccue commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -728,9 +743,20 @@ protected function prepare_status_response( $comment_approved ) {
+		if ( current_user_can( 'unfiltered_html' ) && isset( $prepared_comment['comment_content'] ) ) {
+			$prepared_comment['comment_content'] = wp_filter_post_kses( $prepared_comment['comment_content'] );
+		} elseif ( isset( $prepared_comment['comment_content'] ) ) {
+			$prepared_comment['comment_content'] = wp_filter_kses( $prepared_comment['comment_content'] );
```
Think this should be `wp_kses`, and the above should be `wp_kses_post`

rachelbaker replied:
Doh!

rachelbaker replied:
@rmccue Turns out the problem stems from authenticating as an `admin` level user and then updating a comment for a subscriber.  I don't have easy access to a PHP 5.2 or 5.3 machine to debug this fully.  Any thoughts?

See https://travis-ci.org/WP-API/WP-API/builds/165834410 which finally passes

rachelbaker committed:
Correct sanitization kses functions and re-add failing update content test

rachelbaker committed:
Switch to `wp_filter_kses()` in an attempt to avoid having to pass global allowed tags

rachelbaker committed:
Pass current filter into wp_kses

rachelbaker committed:
Add back string type checking to content values

rachelbaker committed:
Test theory that the problem is caused by and admin updating a comment as a different user

rachelbaker committed:
Merge branch 'develop' into require-comment-content

rachelbaker committed:
Remove conditional check on the `unfiltered_html` cap for discussion

Use the `wp_filter_kses` sanitization function for all new comments

rachelbaker committed:
Revert 6e1f9bac142d8c8dd4293466f47b3d042b190307

This was just a test for debugging why unit tests where failing

rachelbaker commented:
@WP-API/amigos Sorry for the messy commit history here, but I have this ready for review.  I updated the [PR description above](https://github.com/WP-API/WP-API/pull/2741#issue-179942623) to include an accurate summary of the changes here.

kadamwhite reviewed (approved):
I wouldn't use my review as the authoritative go-ahead but the patch looks good to my eyes!

rachelbaker commented:
> I wouldn't use my review as the authoritative go-ahead but the patch looks good to my eyes!

@kadamwhite I just needed a final set of eyes since @rmccue already gave the authoritative go-ahead.  Merged, to keep us moving.
